### PR TITLE
[Fix #81] No negative weeks

### LINKED ISSUE #81: Totally random result?
Between January 31 and March 1 there should be exactly 1 month, or 1 month and 1 day if considering distance rather than interval, but I get this:

    irb(main):008:0> distance_of_time_in_words(Date.new(2014,1,31), Date.new(2014,3,1), locale: :en)
    => "1 month, -1 weeks, and 5 days"

This seems totally off. What's going on?

### PR DESCRIPTION
**Pull request by ivanovaleksey**
It is very dirty hack...
Maybe I can find better solution.

### CODE REVIEW

File: lib/dotiw/time_hash.rb [hunk 1]
```diff
@@ -114,10 +114,17 @@ def build_years_months_weeks_days
...
         days += days_in_month % 7
+
         if days >= 7
           days -= 7
           weeks += 1
         end
+
+        if weeks < 0
+          months -= 1
+          weeks += 5
```
Comment (ivanovaleksey) [inline lib/dotiw/time_hash.rb]: Here `weeks` can only equal to `-1` (as I can see). 
So, we should decrement `months` and `-1` should became `4`.

File: lib/dotiw/time_hash.rb [hunk 2]
```diff
@@ -114,10 +114,17 @@ def build_years_months_weeks_days
...
+          days -= (7 - 3)
```
Comment (ivanovaleksey) [inline lib/dotiw/time_hash.rb]: `days` can equal `5` or `6`. In case of `5` it should be `1` and in case of `6` it should be `2`.

**Comment by bovender** [comment]
@ivanovaleksey, thanks for putting so much effort into this. I've started reading the code myself and attempt to understand the algorithm, but the [`build_years_months_weeks_days`](https://github.com/radar/dotiw/blob/master/lib/dotiw/time_hash.rb#L93) method still leaves me a bit puzzled.

The line that causes the negative week numbers in the first place is

    days = largest.day - smallest.day

With a start date `2014-01-31` and a `2014-03-01`, `days` will be `-30` (i.e., the result of 1-31).

The following line `weeks, days = days.divmod(7)` will then produce negative weeks.

Wouldn't it be best to compute the days in a way similar to the line above

    months = (largest.year - smallest.year) * 12 + (largest.month - smallest.month)

i.e.

    months = (largest.month - smallest.month) * days_in_month + (largest.day - smallest.day)

?

Of course, `days_in_month` is highly variable depending on the months involved.

**Comment by ivanovaleksey** [comment]
@bovender do you mean `days = ...` in the last code snippet, right?
I was thinking about changing the whole `#build_years_months_weeks_days` but I could not find a solution for all cases. That is why I only represent this dirty approach.

In your idea `days_in_month` seems to be a huge problem. I don't see how it could be implemented without any additional hacks.

**Comment by radar** [comment]
Closing because it is old.

**Comment by dblock** [comment]
This is actually a working fix for #75 and #81. I read the comment above and tried the suggestion from @bovender and couldn't make it work. I also have more tests in https://github.com/dblock/distance_of_time_in_words/tree/issue-75.

Will merge after #105.

**Comment by dblock** [comment]
Merged in https://github.com/radar/distance_of_time_in_words/pull/108

**Comment by ivanovaleksey** [comment]
thanks @dblock

**Pull request closed without merging**